pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,245 @@
+package nodelabeller
+
+import (
+	goerror "errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/utils/pointer"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController     *device_manager.DeviceController
+	clientset         kubecli.KubevirtClient
+	namespace         string
+	virtLauncherImage string
+	pullSecret        string
+	host              string
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, clientset kubecli.KubevirtClient, namespace, virtLauncherImage, pullSecret, host string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		clientset:         clientset,
+		namespace:         namespace,
+		virtLauncherImage: virtLauncherImage,
+		pullSecret:        pullSecret,
+		host:              host,
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(stop chan struct{}) error {
+	logger := log.DefaultLogger()
+	logger.Infof("node-labeller is running")
+	kvmPresent := false
+	retries := 0
+	maxRetries := 30
+
+	// check if kvm device is available
+	for {
+		if n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+			kvmPresent = true
+			break
+		}
+
+		select {
+		case <-stop:
+			return goerror.New(fmt.Sprintf("Node-labeller has stopped"))
+		default:
+			retries++
+			if retries > maxRetries {
+				break
+			}
+			time.Sleep(time.Minute)
+		}
+
+	}
+
+	if !kvmPresent {
+		return goerror.New(fmt.Sprintf("Node-labeller cannot work without KVM device."))
+	}
+
+	//create virt-launcher pod to get domcapabilities + info about cpu models
+	pod, err := n.clientset.CoreV1().Pods(n.namespace).Create(n.createVirtLauncherPod())
```

Comment:
I'd rather virt-handler didn't launch pods. I think there's actually a simpler way of handling this though that requires even less effort.

If we need to run a set of containers to discover information about the node for us when virt-handler runs, we can use init-containers [1]. The nice thing about init containers is that they are guaranteed to run before virt-handler can even start, so you'll know all the discovery containers have completed before your virt-handler logic executes to label the node.

you'd just need to add the init containers to the `NewHandlerDaemonSet` function in `pkg/virt-operator/creation/components/deployments.go`

make sense?

1. https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#understanding-init-containers

Reply:
I know what you mean. We are using init containers in node-labeller in ssp operator. In this case it will not work, because virt-launcher needs 
```resources:
            requests:
              devices.kubevirt.io/kvm: "1"
            limits:
              devices.kubevirt.io/kvm: "1"
```
and that device is available after virt-handler is ready.

Reply:
umm, i see. This is a bit of an issue. If we're going to launch a pod to dump the information the labeller needs, the controller doing that needs to be working at the cluster level and not the node level (virt-handler is at node level). 

A new virt-controller reconcile loop dedicated to executing this label job on new nodes that virt-handler starts on would be an option. Another option would be to figure out a way to expose kvm to virt-handler's pod in a way that doesn't using the device plugin. Since virt-handler is privileged already, that might be acceptable.

Reply:
If we can install libvirt and expose kvm to virt-handler, then we don't need to spawn virt-launcher pod and labeller can take all info directly from container.

Reply:
@ksimon1 I think we agreed on init-containers though. You can just use the virt-launcher image as base for the init contianers in virt-handler. The init containers just dump the info in json or any other format and virt-handler does the labeling.

Reply:
For completeness, as discussed offline: Making the init container privileged and using something like

```
if [ ! -e /dev/kvm ]; then
   mknod /dev/kvm c 10 $(grep '\<kvm\>' /proc/misc | cut -f 1 -d' ')
fi
```

allows you to run the init container as part of virt-handler without having to request the kvm device through a device plugin.

Reply:
@rmohr this is the command i had been missing all the time. With this I was able to move virt-launcher to init containers.